### PR TITLE
Add VerifyChannelUpgradeError and tests

### PR DESCRIPTION
**Pull request by chatton**
## Description

part of: #1609

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [x] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Review by damiannolan — APPROVED** [review]
This looks pretty much good to go to me. I just left some small nits/suggestions.
Excellent work, very clean ❤️

File: modules/core/03-connection/keeper/verify_test.go [hunk 1]
```diff
@@ -682,6 +682,178 @@ func (suite *KeeperTestSuite) TestVerifyNextSequenceRecv() {
...
+		suite.Run(tc.name, func() {
+			suite.SetupTest() // reset
+
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.Setup(path)
+
+			errorReceipt = channeltypes.ErrorReceipt{
+				Sequence: 1,
+				Error:    "error message",
+			}
```
[the remark this answers is not in the document]
Comment (chatton) [reply modules/core/03-connection/keeper/verify_test.go:834]: yes! Nice suggestion

File: modules/core/03-connection/keeper/verify_test.go [hunk 2]
```diff
@@ -867,6 +867,88 @@ func (suite *KeeperTestSuite) TestVerifyUpgradeTimeout() {
...
+		tc := tc
+
+		suite.Run(tc.name, func() {
+			suite.SetupTest() // reset
+
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.Setup(path)
+
+			errorReceipt = channeltypes.NewErrorReceipt(1, channeltypes.ErrInvalidChannel)
+			suite.chainA.GetSimApp().IBCKeeper.ChannelKeeper.SetUpgradeErrorReceipt(suite.chainA.GetContext(), path.EndpointA.ChannelConfig.PortID, path.EndpointA.ChannelID, errorReceipt)
```
Comment (AdityaSripal) [inline modules/core/03-connection/keeper/verify_test.go:931]: Could put this in malleate for all cases except key doesn't exist.

File: testing/chain.go [hunk 3]
```diff
@@ -589,3 +589,10 @@ func (chain *TestChain) GetChannelCapability(portID, channelID string) *capabili
 func (chain *TestChain) GetTimeoutHeight() clienttypes.Height {
 	return clienttypes.NewHeight(clienttypes.ParseChainID(chain.ChainID), uint64(chain.GetContext().BlockHeight())+100)
 }
+
+// DeleteKey deletes the specified key from the ibc store.
+func (chain *TestChain) DeleteKey(key []byte) {
```
Comment (AdityaSripal) [inline testing/chain.go:594]: Don't really feel this helper function needs to exist.

Either do this inside the one test case you need it. Or explicitly do the set error receipt in the other test cases,

**Review by AdityaSripal — APPROVED** [review]
Nice work!! Only have a nit on testing

File: testing/chain.go (hunk 3 above)
Comment (chatton) [reply testing/chain.go:594]: Sorry missed this, can remove the helper in a follow up if you don't think it's worth having 👍

**Pull request merged**